src/test/java/fr/sncf/osrd/Helpers.java

```diff
@@ -336,4 +323,80 @@ public static void makeFunctionEvent(Simulation sim, double time, Procedure func
+    public static SortedDoubleMap getTimePerPosition(Iterable<TimelineEvent> events) {
+        var res = new SortedDoubleMap();
+        for (var event : events) {
+            if (event instanceof TrainReachesActionPoint) {
+                var trainReachesActionPoint = (TrainReachesActionPoint) event;
+                for (var update : trainReachesActionPoint.trainStateChange.positionUpdates)
+                    res.put(update.pathPosition, update.time);
+            }
+        }
+        return res;
+    }
+
+    /** Throws an error if the (interpolated) time per position differ too much */
+    public static void assertSameSpeedPerPosition(Iterable<TimelineEvent> eventsExpected,
+                                                  Iterable<TimelineEvent> events) {
+        assertSameSpeedPerPositionBetween(eventsExpected, events, 0, Double.POSITIVE_INFINITY, 1);
+    }
+
+    /** Throws an error if the (interpolated) time per position differ too much between begin and end
+     * The expected speeds are scaled by expectedScale */
+    public static void assertSameSpeedPerPositionBetween(Iterable<TimelineEvent> eventsExpected,
+                                                         Iterable<TimelineEvent> events,
+                                                         double begin,
+                                                         double end,
+                                                         double expectedScale) {
+        var expectedSpeedPerPosition = getSpeedPerPosition(eventsExpected);
+        var speedPerPosition = getSpeedPerPosition(events);
+        end = Double.min(end, expectedSpeedPerPosition.lastKey());
+        begin = Double.max(expectedSpeedPerPosition.firstKey(), begin);
+        for (double t = begin; t < end; t += 1) {
+            var expected = expectedSpeedPerPosition.interpolate(t) * expectedScale;
+            var result = speedPerPosition.interpolate(t);
+            assertEquals(expected, result, expected * 0.01);
+        }
+    }
+
+    /** Get a map of the speed at each position */
+    public static SortedDoubleMap getSpeedPerPosition(Iterable<TimelineEvent> events) {
+        var res = new SortedDoubleMap();
+        for (var event : events) {
+            if (event instanceof TrainReachesActionPoint) {
+                var trainReachesActionPoint = (TrainReachesActionPoint) event;
+                for (var update : trainReachesActionPoint.trainStateChange.positionUpdates)
+                    res.put(update.pathPosition, update.speed);
+            }
+        }
+        return res;
+    }
+
+    /** Adds an element to an array of route
+     * This function is used to hide away the generic array creation */
+    @SuppressWarnings("unchecked")
+    public static ID<RJSRoute>[] addToArray(ID<RJSRoute>[] array, ID<RJSRoute> val) {
+        var res = (ID<RJSRoute>[]) java.lang.reflect.Array.newInstance(ID.class, array.length + 1);
```

Comment:
java.lang.reflect.Array.newInstance(ID.class, array.length + 1) is the same thing as new ID[array.length + 1];